src/Microsoft.AspNetCore.Blazor/Components/ComponentResolver.cs

```diff
@@ -22,37 +22,49 @@ public static IEnumerable<Type> ResolveComponents(Assembly appAssembly)
         {
             var blazorAssembly = typeof(IComponent).Assembly;
 
-            return EnumerateAssemblies(appAssembly.GetName(), blazorAssembly, new HashSet<Assembly>(new AssemblyComparer()))
+            return EnumerateAssemblies(appAssembly.GetName(), blazorAssembly)
                 .SelectMany(a => a.ExportedTypes)
                 .Where(t => typeof(IComponent).IsAssignableFrom(t));
         }
 
         private static IEnumerable<Assembly> EnumerateAssemblies(
```

Comment:
I'm unsure about this updated logic. Does it correctly handle a dependency graph like the following?

```
AppAssembly
 - Direct1
   - Indirect
     - Blazor
 - Direct2
   - Indirect
     - Blazor
```

If I'm following the algorithm correctly, it looks to me like the second time it sees `Indirect`, it will skip it because it's already in `visited`. Then the candidate count will not be increased, which means that it will not recognize that `Direct2` transitively references `Blazor`, so `Direct2` will not be in the output list even though it should be.

If I'm wrong about this, could you let me know what I'm misunderstanding?

If I'm right about this, maybe this approach is overcomplicated, unless we are going to add a set of test cases for it as well.

TBH I'd be satisfied with the simpler rule of "consider all transitively-referenced assemblies to be candidates, but don't recurse into any assemblies whose name starts `System.*` or is mscorlib".

Reply:
Assembly scanning destroys startup time. Can we consider an approach that does most of the filtering at build time? Is this what is happening here?

Reply:
As long as the scanning skips `mscorlib` plus any `System.*` assemblies and their dependencies (which it doesn't yet, but I'm proposing it should), then the total amount of scanning shouldn't be too bad.

I agree it *would* be better to try to resolve as much as we can at build time, but that's quite a bit of extra machinery. It's likely that we'll change how this stuff works again when we implement routing enhancements (to mount external assembly components at a certain path in the URL space) and lazy-loading. So I'd be fine with a limited amount of runtime scanning on startup until we get closer to a complete design with those other features.

Reply:
I will add some tests here just to make sure. If we want to prevent assembly scanning we can just do the same thing Razor does and put an assembly level attribute on the assembly that indicates what the components in the assembly are, but as @SteveSandersonMS said. That can wait until later.

Reply:
Just to drop my 2 cents. As @davidfowl suggested, in Orleans we dropped assembly scanning and instead we adopted the concept of _Application Parts_ from ASP.Net. At startup time, the user register with an `IAppPartsManager` all the assemblies that we will need to look for a given type (in this case the router types). 

Something like this:

```
...
.ConfigureApplicationParts(parts => parts.AddApplicationPart(typeof(ValueGrain).Assembly).WithReferences())
...
```

An assembly attribute is not good as it hides from the user the real purpose of that. Being explicit is always the best option to avoid confusion IMHO.